### ISSUE TITLE
Recommended monitoring tools/management

### ISSUE THREAD

**Issue by KG7QIN** [rocky-linux/infrastructure#22]
I ran across the Infrastructure discussion on Discourse.  Lots of good suggestions there for monitoring tools.   

I'd like to offer a few that are used in the back end of allstarlink.org:

LibreNMS - used for monitoring servers and services.  Pretty configurable and has lots of options for alerts.  It can even feed its data into InfluxDB or Prometheus for use in Grafana.

Graylog - an open source SIEM that takes parts of the ELK stack and can use other things as well to ingest logs and store them in an Elasticsearch back end.  Lots of options for things too like pipelines, streams, events/alerts and notifications.  Can be clustered together as well easily.  The community edition offers quite a few options and it has a good API to query for pulling information if you don't want to query Elasticsearch directly.  

SaltStack - While Ansible is more widely known, SaltStack is a good powerful replacement that also does orchestration and has an event management feature (reactor) where a node (minion) can report back and have it kick off automation to remedy an issue. 

Hope this helps.

**Comment by nazunalika** [comment]
Thank you for the input!

For the configuration management, we have already decided on AWX/Ansible alongside other tooling. We are currently emulating as close as possible to the CentOS and Fedora projects respectively, who use ansible to stand-up and operate their infrastructure. Currently, our contributors and some of leadership have more experience in Ansible, and thus it was the chosen tool to handle infrastructure and beyond. This isn't to say saltstack is bad (it isn't, I used to maintain it at another job in a past life), we just made the choice to stick with Ansible as we have a lot of resources to bring it together.

As for the others, I do not believe we have come to any consensus on those things, but they are good suggestions just like in the forum thread. I actually like graylog and have used it before, so it does have my vote when it comes time to talk about it. 

I will leave this issue open for further discussion with the other leaders and contributors.

Related discourse forum: https://forums.rockylinux.org/t/what-servers-services-do-we-need-to-bootstrap-ourselves/65/

**Comment by Gorian** [comment]
I've deployed graylog a number of times and like it. I've contributed to LibreNMS as well and the rest of the LibreNMS team is pretty great in addition to having a good product - both are good choices

**Comment by y0rune** [comment]
For infrastructure I'd recommend the Zabbix.

**Comment by chriscowley** [comment]
Personally I would run as far away from anything that even resembles Nagios as I possibly can. Zabbix marginally better, but not by much. My suggestion would be the following:

Monitoring - Prometheus
Logging - ELK

Prometheus implies a certain amount of other tools:

- Consul (or some other form of Service Discovery)
- Prometheus itself
- The exporters
- Grafana for visualisation
- Long-term storage - Thanos seems to have the community mind share
- AlertManager connected to something to send the alerts - perhaps approach Pagerduty, or similar, for sponsorship

ELK, I am not an expert so I will bow out on that subject for now

**Comment by scepticozymandias** [comment]
I would also suggest staying away from anything that is Nagios related. I think this product might be good enough for some existing projects. I would also vote for Prometheus or Zabbix.

**Comment by y0rune** [comment]
I used the Prometheus for the IOT but for the server maintenance. We can also use zabbix for information or creating a incident in the OpenProject. Or even call system to our OutOfOffice Hours.

**Comment by derekmpage** [comment]
> 
> 
> Personally I would run as far away from anything that even resembles Nagios as I possibly can. Zabbix marginally better, but not by much. My suggestion would be the following:
> 
> Monitoring - Prometheus
> Logging - ELK
> 
> Prometheus implies a certain amount of other tools:
> 
>     * Consul (or some other form of Service Discovery)
> 
>     * Prometheus itself
> 
>     * The exporters
> 
>     * Grafana for visualisation
> 
>     * Long-term storage - Thanos seems to have the community mind share
> 
>     * AlertManager connected to something to send the alerts - perhaps approach Pagerduty, or similar, for sponsorship
> 
> 
> ELK, I am not an expert so I will bow out on that subject for now



I agree with all of this.

We just implemented Thanos at my $dayjob backed by minio. Pretty simple stuff.

We are also an ELK shop so between @NeilHanlon and I we have plenty of experience.

**Comment by chriscowley** [comment]
I have personally used Thanos on GCP (using GCS as storage), but seeing as it was announced today that we will be using AWS, I would happily leave you guys to configure that side of it seeing as you have already set it up with S3/Minio.

I think the MVP does not need Thanos, the basic Prometheus TSDB will be fine initially, but it should definitely be added with in a few weeks.

**Comment by kiransubash** [comment]
For log aggregation, another option is to look at Loki. It integrates well along with Prometheus and Grafana, and with the newer versions, the need for separate storage backends is gone. It will happily use the S3/Minio backed storage, and scale.

**Comment by chriscowley** [comment]
@derekmpage SNMP monitoring will be over my dead body. I was on call last week and more than half the alerts that woke me up were simply because `snmpd` had silently stopped responding.

Yes, if we have any switches/routers to monitor, SNMP will perhaps be necessary, but it should certainly not be the core.

**Comment by chriscowley** [comment]
@kiransubash yep, I hear what you're saying about Loki. I had not thought of it because I associate it with Kubernetes, but you are right. Also it has the advantage that it can be mixed into the same Grafana instance.

**Comment by NeilHanlon** [comment]
> @derekmpage SNMP monitoring will be over my dead body. I was on call last week and more than half the alerts that woke me up were simply because `snmpd` had silently stopped responding.
> 
> Yes, if we have any switches/routers to monitor, SNMP will perhaps be necessary, but it should certainly not be the core.

Don't worry. Snmpd changes the disk mibs on us if it restarts. So disks page if Snmpd restarts.

**Comment by derekmpage** [comment]
> 
> 
> > @derekmpage SNMP monitoring will be over my dead body. I was on call last week and more than half the alerts that woke me up were simply because `snmpd` had silently stopped responding.
> > Yes, if we have any switches/routers to monitor, SNMP will perhaps be necessary, but it should certainly not be the core.
> 
> Don't worry. Snmpd changes the disk mibs on us if it restarts. So disks page if Snmpd restarts.

Yes, and if you are monitor disks and you have a hung NFS mount SNMP walk or gets of that tree will just hang.

Of course we will not be using any NFS right? :)

**Comment by jkwong** [comment]
There's an open source alternative to PagerDuty called Cabot (https://github.com/arachnys/cabot).  Another open source SIEM solution is Wazuh (https://github.com/wazuh/wazuh), which is a fork of OSSEC and runs on top of the Elastic Stack.

**Comment by RaceAap** [comment]
Why use Prometeus and ELK, and not only ELK ?
ELK is more than only a Logging tool, it has all capabilites for Monitoring and Logging build-in

**Comment by NeilHanlon** [comment]
> Why use Prometeus and ELK, and not only ELK ?
> ELK is more than only a Logging tool, it has all capabilites for Monitoring and Logging build-in
> 

Elk kinda...

Sucks. :/

**Comment by mddeff** [comment]
Have we defined what the "job" is?  Just basic health and monitoring of the OS stacks for each infra service? Some basic service health checks (http/https, certs, etc.)?

If we're looking for something more "traditional" (monitoring server that issues the checks) but want to avoid nagios, look at icinga2.  It leverages the massive database of nagios check plugins and formatting with a completely rewritten backend with a clustering/agent model.  All configurable from the web interface, but gives you all the plumbing to deploy it as infra-as-code.

Or are we looking at something more modern?  "Cloud Native" if you will.

- Grafana/prometheus/loki with telegraf for local host metrics that aren't supported with prometheus.
- Don't discount ELK.  It's a good solution with lots of capability included via the beats.
  - Though, unless you have x-pack, a lot of the security/auditing one would need in production needs to be painted on the top with the likes of gluu/keycloak.  Not baked in.
- Avoid Splunk unless you can afford a full time Splunk engineer...

Also, if you're looking for something to do metric transport/aggregration/manipulation (or really any dataflow work), check out [Apache Nifi](https://nifi.apache.org/)

**Comment by derekmpage** [comment]
ELK handles and stores time series data very poorly.

Prometheus cans grafana have already defined dashboard alerts rules and
mixins. ;)

**Comment by chriscowley** [comment]
@mddeff as I understand the needs are:

- collect and alert on infrastructure (hosts and services)
- Open source
- vendor neutral
- simple to get started, but not restrictive as we grow
- something that volunteers will be either familiar with and/or willing to learn

I have added my personal specification: If it has anything even vaguely to do with Nagios, I will release the flying-monkeys.

**Comment by chriscowley** [comment]
Also @mddeff, our most righteous infrastructure manager (@NeilHanlon) has spoken on the subject of ELK for metrics:

> Elk kinda...
> 
> Sucks. :/

**Comment by mddeff** [comment]
Got it, ELK is out.  So it is written, so it shall be (not) done!  Looks like (based off of your #14944 pull request and comments here) that Prometheus has been decided on.  Happy to help with implementation when the time comes.

**Comment by RaceAap** [comment]
> Also @mddeff, our most righteous infrastructure manager (@NeilHanlon) has spoken on the subject of ELK for metrics:
> 
> > Elk kinda...
> > Sucks. :/
> 
> And I agree, although it is still better than Nagios

You can't compare Nagios (Core) with ELK. It are two very different tools.

**Comment by kiransubash** [comment]
@RaceAap Loki serves central log storage very well, and definitely reduces the bulk in storage volume. Especially if we manage the tags well, it will be much less than ELK, and searching will be much faster. Since we have prometheus already, the syntax is very identical to query, and integration in Grafana where you can switch from metrics to logs is very appealing.

**Comment by chriscowley** [comment]
@RaceAap ELK will definitely be considered for the **logging** solution. I am not even trying to address that here, this is for metrics and alerting. They are not at all the same problem.

As others have said, Loki is perhaps a good solution for log aggregation. It will integrate nicely with Grafana, but I have not got enough time it to make an informed decision.

**Comment by mddeff** [comment]
Given that the [plan forward for logging/auditing seems to be ELK/ElasticStack](https://github.com/rocky-linux/infrastructure/issues/14946#issuecomment-746896771), I just have to ask one more time.  Are we sure we dont want to use metric/heartbeat for monitoring through ELK?

@chriscowley and @NeilHanlon,  clearly you feel that ES is a bad solution for monitoring, could you guys elaborate more on:

> Elk kinda...
> 
> Sucks. :/
> 
> That is to say it's not the right tool for the job.

I'm about to roll out Beats at work to thousands of endpoints for my day job using {metric,file,heart,journal}beat.  There are several other teams at my job that are already sending ~50-100MBPS+ of system and app monitoring/telemetry to the corporate ES stack.

That being said, I'm just a consumer of the "ES-as-a-service".  We've got a whole team of 10+ guys who do nothing but manage the ES stacks on our various production networks.  I have no real world experience of what it takes to run ES at that scale in production.

As a user, its awesome! I fire up the beats, point them at the ES/Logstash ingest point, and I'm done!  Plenty of built in dashboards, alerting, etc. setup in Kibana by the beats applications.

As I said in #14964:
> Unless there's a good technical reason to go with something else. That technical reason just needs to outweigh the very real costs (time, money, operational complexity) in running separate logging/monitoring solutions.

Is ES so bad for *monitoring* that it really is worth doubling up on infrastructure?  I think we all need to remember that at the end of the day, except for a few of us, this is a volunteer based project.  By the nature of this project's genesis, we're eschewing corporate resources (time, money, mostly time) that would afford us these kind of "expensive" architectural decisions.

If ES/ELK truly is a hot dumpster fire for monitoring but good for logging, please, share how! Like I said, I have *no operational experience running a production grade ES cluster*.  If nothing else, I'd love to learn so that I don't make this mistake on other projects both professional and personal.  Thanks!

**Comment by derekmpage** [comment]
> 
> 
> Given that the [plan forward for logging/auditing seems to be ELK/ElasticStack](https://github.com/rocky-linux/infrastructure/issues/14946#issuecomment-746896771), I just have to ask one more time. Are we sure we dont want to use metric/heartbeat for monitoring through ELK?
> 
> @chriscowley and @NeilHanlon, clearly you feel that ES is a bad solution for monitoring, could you guys elaborate more on:
> 
> > Elk kinda...
> > Sucks. :/
> > That is to say it's not the right tool for the job.
> 
> I'm about to roll out Beats at work to thousands of endpoints for my day job using {metric,file,heart,journal}beat. There are several other teams at my job that are already sending ~50-100MBPS+ of system and app monitoring/telemetry to the corporate ES stack.
> 
> That being said, I'm just a consumer of the "ES-as-a-service". We've got a whole team of 10+ guys who do nothing but manage the ES stacks on our various production networks. I have no real world experience of what it takes to run ES at that scale in production.
> 
> As a user, its awesome! I fire up the beats, point them at the ES/Logstash ingest point, and I'm done! Plenty of built in dashboards, alerting, etc. setup in Kibana by the beats applications.
> 
> As I said in #14964:
> 
> > Unless there's a good technical reason to go with something else. That technical reason just needs to outweigh the very real costs (time, money, operational complexity) in running separate logging/monitoring solutions.
> 
> Is ES so bad for _monitoring_ that it really is worth doubling up on infrastructure? I think we all need to remember that at the end of the day, except for a few of us, this is a volunteer based project. By the nature of this project's genesis, we're eschewing corporate resources (time, money, mostly time) that would afford us these kind of "expensive" architectural decisions.
> 
> If ES/ELK truly is a hot dumpster fire for monitoring but good for logging, please, share how! Like I said, I have _no operational experience running a production grade ES cluster_. If nothing else, I'd love to learn so that I don't make this mistake on other projects both professional and personal. Thanks!



I will answer the last two questions IMHO from my experiences from $dayjob.

> Is ES so bad for _monitoring_ that it really is worth doubling up on infrastructure? I think we all need to remember that at the end of the day, except for a few of us, this is a volunteer based project. By the nature of this project's genesis, we're eschewing corporate resources (time, money, mostly time) that would afford us these kind of "expensive" architectural decisions.

ES can be a dog. Run by java it requires a massive amount of resources and Garbage Collection tuning; which despite volunteers from the community this is still a lot of over head / time / money.

Elastic search runs best on a cluster with "like" indices, so for log ingestion the basic index is simple and alike.

[ "log source", 
 "log message" ]

When you end up with various data sources as you do in the metrics world these indecies tend vary a lot requiring physical separation of clusters and therefor more resources and more overhead.

If you have every played with logstash you will understand the resource requirements for GROK parsing logs at an enormous rate is massive. Hopefully basic logging/auditing rates will be low.

Beats written in GO are usually fast and help to fix some of the infrastructure overhead, but sometimes you just end up with so many moving parts to support all the different types of ingestion. More moving parts, more to monitor, more to fix and baby sit.

Prom does metrics perfect and I just don't understand why we would do anything else. OpenMetics is the new standard and most things are playing along. 

Not to mention there is are already grafana dashboards/rules/alerts for pretty much anything you can imagine, which I don't find this much with ELK metrics world.

So to answer your second last question.

Yes, "ELK is a hot dumpster for monitoring but good for logging."

P.S. One thing our ELK does not do at my $dayjob is "alert", mainly because elastalert "sucks". 

Again, this is just my 2cents from my own experiences.

**Comment by chriscowley** [comment]
@mddeff I'm not going to repeat what @derekmpage has said. He articulated it far better than can with only 1 mug tea in my system today.

I'll add that ES is designed for text search, not as a TSDB. Using as such is not using in its best light. 

Also the only ES clusters under my care that actually ENJOY administering are those that run under K8s (specifically using the relatively ne elastic operator). For now we are not planning a K8s cluster (one day @NeilHanlon, one day). That is a personal thing I know, but volunteers have to want to get their teeth in to these things.

**Comment by RaceAap** [comment]
> If you have every played with logstash you will understand the resource requirements for GROK parsing logs at an enormous rate is massive. Hopefully basic logging/auditing rates will be low.
> 
But Logstash is can be replaced with Fluentd and that has much more options and easier to learn than those grumpy grok queries and it need less resources when parsing the logs too.
So than it would not be ELK but EFK

**Comment by derekmpage** [comment]
> 
> 
> > If you have every played with logstash you will understand the resource requirements for GROK parsing logs at an enormous rate is massive. Hopefully basic logging/auditing rates will be low.
> 
> But Logstash is can be replaced with Fluentd and that has much more options and easier to learn than those grumpy grok queries and it need less resources when parsing the logs too.
> So than it would not be ELK but EFK

Fluentd is only something I have dreamt about and have not had the pleasure of working with it yet.
Good to know it is a viable replacement for the ankle weight that logstash is. Thanks for the tip.

**Comment by mddeff** [comment]
@derekmpage Thank you for the detailed response; all good information to know.  I became all too aware of ELKs massive memory requirements (in comparison to other solutions) when I went to spin it up in one of my k3s clusters at home.
> mainly because elastalert "sucks".
Good to know.

---
> For now we are not planning a K8s cluster (one day @NeilHanlon, one day).

@chriscowley you let me know when that day comes, I'm all about moving stuff to Kube, whatever the flavor.  The fact that its a desired state system soup to nuts (not just some automation thrown on top) makes things so much more repeatable IMHO (and subsequently easier).  And the barrier to entry is so low now a days, its hard to say no to.  But like you've said, walk before we can sprint.

> volunteers have to want to get their teeth in to these things
Another debate we have at my office every, single, day.  Solid copy.

**Comment by derekmpage** [comment]
@mddeff 

> mainly because elastalert "sucks".

I should note, it has been years since I have used it... Things could be better now :)

**Comment by chriscowley** [comment]
> @mddeff 
> 
> > mainly because elastalert "sucks".
> 
> I should note, it has been years since I have used it... Things could be better now :)

It's not really Elastic's core business, so I doubt it

**Comment by RaceAap** [comment]
> > @mddeff
> > > mainly because elastalert "sucks".
> > 
> > 
> > I should note, it has been years since I have used it... Things could be better now :)
> 
> It's not really Elastic's core business, so I doubt it

As of Elasticsearch 7.x the notifications have had a very big change in Kibana and their main focus changed to add monitoring instead of only log analytics.

But why still trying to bash on ELK as for the central logging/auditing ELK is already chosen as the tool to use ?

**Comment by derekmpage** [comment]
Yes, ELK is chosen for logging/auditng tool.

I would not say we are bashing ELK just being cautious of some of it's flaws. But I won't disagree, we are beating a dead horse at this point.

**Comment by RaceAap** [comment]
On the [wiki](https://wiki.rockylinux.org/en/Infrastructure/services) I saw that prometeus is apparently the chosen monitoring tool, or are there still other tools to discuss instead of prometeus?

**Comment by chriscowley** [comment]
In my opinion no, not really. There is debate about the architecture (HA or not in the initial build), but I would call that a new issue.

**Comment by Zaragon-AQ** [comment]
IMO:
Monitoring: prometheus, alertmanager, grafana
Logging: ELK or Open Distro for Elasticsearch https://opendistro.github.io/for-elasticsearch/

**Comment by christophedumont5** [comment]
Monitoring : Prometheus and Nagios
Logging : ELK

**Comment by RaceAap** [comment]
> Monitoring : Prometheus, Grafana and Nagios
> Logging : ELK (graylog ?)

Nagios was already voted for not to be used
And MongoDB is a disaster when it comes to upgrades and keeping it secure, so I vote NO for Graylog.

For Logging is/was an other issue: Log aggregation/archive tools #14946 and that has been handled already so please stop discussing about Logging tools in a thread for Monitoring tools.

**Comment by richard-ostrochovsky** [comment]
My suggestion:

**ZABBIX** (ZBX) as general purpose "umbrella" monitoring, providing mature admin/user frontend with problem view (providing kind of workflow on incidents/problems - acking and closing problems, and many more), dashboards, maps, event correlation, problem tagging, powerful preprocessing, various discovery mechanisms, integrations, ZBX API, scalability using ZBX proxies, etc. etc. etc. which during the years evolved into something, that is designed and ready for IT Operations (ITO) processes ... there are also possiblities of monitored services decomposition (BSM service tree)

of course, there are also many other open-source monitoring tools, that fit more tightly into specific domains, but they are often missing something important for ITO, and often need to be integrated with something else (e.g. ZBX ... or Grafana, but Grafana is more of visualisation/observability, than ITO-workflow platform, that ZBX frontend is, and Grafana has its Enterprise version, and we know, what that means ... ZBX does not have paid "enterprise features", everything is free software, and there is a long term promise, that this model won't change, ZBX vendor is building its profit on providing professional services, not enterprise features)

we can view some of that other monitoring tools as something to be integrated into "umbrella" monitoring, I truly recommend to keep them too (where suitable / reasonable):

- Graylog
- Prometheus
- Grafana

and integrate them with ZBX, because all problems/incidents **shall** appear in one console (ZBX frontend suggested, maybe some more advanced visualizatons in Grafana)

ZBX DB backend: I suggest to use **TimescaleDB** (time-series extension for PostgreSQL)

there is also great community around ZBX, and here is som tasting from what's new in newest ZBX 5.2:
https://www.youtube.com/watch?v=Wq8qUFjPTP8&list=PL0mpo4XDX5eGaACQsHfgeKlBE-AK7pXRs&index=7 (part of playlist from ZABBIX Summit Online 2020)

there were also long time support for CentOS/RHEL, and as CentOS dead-ending, I see there space for narrower cooperation between ZABBIX and Rocky Linux, to be able replace CentOS with Rocky Linux here:
https://www.zabbix.com/documentation/current/manual/installation/install_from_packages/rhel_centos
in case of interest, I personally could help with this process, and also with monitoring infrastructure of Rocky Linux project itself

**Comment by NeilHanlon** [comment]
Re: monitoring

The tl;dr of this is that prometheus is the direction we're going in for monitoring. 

It's been gone back and forth over and over on the forums, github, slack, threads, more threads, pull requests, who knows. At the end of the day prometheus and that stack is what is known best here and what has been decided to be the solution that best meets the needs of our infrastructure. Debating it back and forth will not lead us anywhere.

Prometheus for monitoring.
